### PR TITLE
Adding sub dialogue for formatting of summary tables

### LINKED ISSUE #6770: Formatting table output for mmtable2
We have introduced `mmtable2` into R-Instat to create summary tables. The package works similar to `ggplot2` where layers are added through `+` to enhance or change the table.

There is a vignette [here](https://ianmoran11.github.io/mmtable2/articles/Some-minimal-examples.html) which has three sections:
* Adding headers  - we have done this already in the summary and frequency tables dialogs.
* Combining tables  - looks exciting and I'm sure will be introduced into R-Instat in the future!
* Formatting tables  - what we are interested in for this issue

Formatting tables allows changes to be made to the mmtable2 object using the `gt` package. There is some sample code for formatting tables at the bottom of this comment, but it looks like there are many different areas where we can format the tables.

## Formatting Tables

`cells_format` - this is to format the cells in the tables
`header_format` - this is to format the headers in the tables
`table_format` - this is to format the `mmtable2` object altogether

For these functions 

* The first parameter states what to format
E.g. for `header_format`, which header to format.
E.g.2. for `table_format`, `cells_body` from the `gt` package can be read in to specify which rows or columns or cells to format.

* The second parameter they can take is `style` which specifies how to format it.
E.g. `style = list(cell_text(...))` where `cell_text` is from the `gt` package.

From a quick look, `cell_borders` and `cell_fill` are other parameters that can be read into `style = list(...)`. 

So one task is to explore what, if any, other parameters can be read into the `style = list(...)` function from the `gt` package. A quick glance at the `tab_style` help page in the `gt` package suggests it is only these three, however perhaps there are more that I have missed.

Another task is to then look at what parameters these functions can take.
That is, what parameters `cell_text` can take in R, and similarly so for `cell_fill` and `cell_borders`.

## Titles, Captions, Etc.
The `table_source_note` function is another layer that can be added. This takes arguments from `gt::tab_source_note`, so the parameter this takes is `source_note` with the value as a string of what the caption should say.

From a quick look of the `gt` package, `tab_header` is another layer that can add a title and subtitle to the `mmtable2` object. 

A third task is what else from the `gt` package works as a new layer for an `mmtable2` package. It might be worth looking at the [vignette](https://cran.r-project.org/web/packages/gt/gt.pdf).
From a quick google, https://gt.rstudio.com/articles/intro-creating-gt-tables.html discusses other bits to add (such as footnotes) so it could be worth having a look there. 

## Other Options 
`header_merged_cols()` is another option in the `mmtable2` function to edit the headers
Are there any other options?

Finally, we then want to take all of these bits and plan a dialog or sub-dialog to format the `mmtable2` object in R-Instat.

## R Code
```
#remotes::install_github("ianmoran11/mmtable2")

library(gapminder)
library(dplyr)
library(tidyr)
library(stringr)
library(purrr)
library(gt)
library(mmtable2)

gapminder_mm %>% 
  filter(var != "Life expectancy") %>% 
  mmtable(cells = value) +
  header_top(year) +
  header_left(country) +
  header_top_left(var)  +
  header_left_top(continent)  +
  cells_format(cell_predicate = T, style = list(cell_text(align = "right"))) +
  header_format(header = year, style = list(cell_text(align = "right"))) +
  header_format("all_cols", style = list(cell_text(weight = "bolder"))) +
  header_format("all_rows", style = list(cell_text(weight = "bolder"))) +
  table_format(
    locations = cells_body(rows = c(1,3,5,7,9,11)),
    style = list(cell_borders(sides = "top",color = "grey"))) + 
  table_source_note(source_note = "Excerpt of the Gapminder data on life expectancy, GDP per capita, and population by country." )
```

(As a side note, I can see some functions called `cols_hide` and bits in the `gt` package, perhaps that would be an interesting avenue to explore later for editing mmtable2 objects. However, that is not to do with formatting).

### PR DESCRIPTION
**Pull request by Ivanluv**
fixes #6770
This is still in progress

### CODE REVIEW

**Comment by lloyddewit** [comment]
@Ivanluv thanks for taking on this issue.
If a PR is not yet ready for review, then you can set it to draft (see the 'convert to draft' link in top-right of the screen).
image

**Comment by Ivanluv** [comment]
> @Ivanluv thanks for taking on this issue. If a PR is not yet ready for review, then you can set it to draft (see the 'convert to draft' link in top-right of the screen). image

Thanks for this

**Comment by lloyddewit** [comment]
@Ivanluv Please could you clarify the status of this PR? thanks

**Comment by Ivanluv** [comment]
@africanmathsinitiative/developers  this is ready for review. @lilyclements  could you take a look too

**Comment by lloyddewit** [comment]
@Ivanluv Thank you! this looks like a significant PR.
@rdstern Please could you do an initial test?
@N-thony Please could you do an initial review?
Thanks

**Review by N-thony — CHANGES_REQUESTED** [review]
Could you also add at the top one line space before the import?

' R- Instat
' Copyright (C) 2015-2017
'
' This program is free software: you can redistribute it and/or modify
' it under the terms of the GNU General Public License as published by
' the Free Software Foundation, either version 3 of the License, or
' (at your option) any later version.
'
' This program is distributed in the hope that it will be useful,
' but WITHOUT ANY WARRANTY; without even the implied warranty of
' MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
' GNU General Public License for more details.
'
' You should have received a copy of the GNU General Public License 
' along with this program.  If not, see <http://www.gnu.org/licenses/>.

File: instat/sdgFormatSummaryTables.vb [hunk 1]
```diff
@@ -1,3 +1,552 @@
-﻿Public Class sdgFormatSummaryTables
+﻿Imports instat.Translations
+Public Class sdgFormatSummaryTables
+    Private clsTableTitleFunction, clsTabFootnoteTitleFunction, clsTableSourcenoteFunction, clsFootnoteCellFunction,
+        clsCellTextFunction, clsCellBorderFunction, clsCellFillFunction, clsHeaderFormatFunction, clsFootnoteTitleLocationFunction,
+        clsTabOptionsFunction, clsBorderWeightPxFunction, clsFootnoteSubtitleLocationFunction, clsTabFootnoteSubtitleFunction, clsFootnoteCellBodyFunction,
+        clsStyleListFunction, clsStubHeadFunction As New RFunction
+    Private clsPipeOperator, clsMutableOperator, clsJoiningOperator As New ROperator
+    Private bControlsInitialised = False
```
Comment (N-thony) [inline instat/sdgFormatSummaryTables.vb:8]: As Boolean?

**Comment by rdstern** [comment]
As @lloyddewit has said, this is a major new feature.  This sub-dialogue is starting (I assume) a similar "system" to the sub-dialogue for the graphs.
a) So please could it also be added to the Frequency Tables dialogue at the same time.  I assume it will be identical there, and this will help in testing the features.  I assume it will replace the disabled Options button?
b) So I note first a few small items.  I would like to merge it as soon as possible, but assume improvements will be on-going work for a bit of time yet.
c) On the title tab it gives me an error when I tick anything, but leave it all empty.  That's for any of the 3 categories.  Could it ignore the tick if it is then empty? Or capture that in a more friendly way?  It doesn't seem to need everything completed, but does need something.
d) The footnote control is quite small.  There is space to make them larger without changing the size of the sub-dialogue.
e) Add sourcenote seems a nice command name, but perhaps could be Add Source on the dialogue.
f) For the footnote I see we can only have a single footnote?  How could we have (say) 2 footnotes?
g) On the main dialogue, it needs to become a bit shorter.  I wonder about:
1) Moving the Summary button to above the Summary control.
2) Making the Summary multiple receiver a bit smaller.
3) Move the Store Output to the bottom - where the summaries button is now?
4) Move the Margin Name up a git, so on the same level as the margin checkbox
5) Then the 3 margin options can go across.
h) Will there be more options on the titles tab, for example font, sixe?
i) The header is either for rows or columns?  Should there be 2 tables, so each can be specified?
j) Most are set to NULL as the default, but the 2 colour ones are set initially to a (confusing) number - though NULL is available.
k) On the Table tab I am not sure what to expect.  Can Background Colour be without the capital G.  I Tried to make it yellow and it coloured the title and footnote.  Is that what it was meant to do?  Can I get the sort of formatting colours that Excel gives, with alternate rows, etc?

**Comment by Ivanluv** [comment]
@rdstern  could you have a look again
These are the changes I have added:

1. Added the format table sub dialogue to the frequency table
2. Fixed the error On the title tab when  anything is ticked  while leave it all empty. This I have fixed by  having the footnote ,title and  source note functions only when the labels are provided
3. Made design changes to the summary dialogue and minor changes to the sub dialogue
4. Added the option for the second footnote
5. Added the font and size options for the title

> 5\. j) Most are set to NULL as the default, but the 2 colour ones are set initially to a (confusing) number - though NULL is available

Those numbers are the R defaults values for those colours



> 5\. k) On the Table tab I am not sure what to expect.  Can Background Colour be without the capital G.  I Tried to make it yellow and it coloured the title and footnote.  Is that what it was meant to do?  Can I get the sort of formatting colours that Excel gives, with alternate rows, etc?

I might need help with these

File: instat/sdgFormatSummaryTables.vb [hunk 2]
```diff
@@ -1,3 +1,552 @@
...
+        dctTextSize.Add("small", Chr(34) & "small" & Chr(34))
+        dctTextSize.Add("medium", Chr(34) & "medium" & Chr(34))
+        dctTextSize.Add("large", Chr(34) & "large" & Chr(34))
+        dctTextSize.Add("x-large", Chr(34) & "x-large" & Chr(34))
+        dctTextSize.Add("xx-large", Chr(34) & "xx-large" & Chr(34))
+        ucrInputStyleTextSize.SetItems(dctTextSize)
+        ucrInputStyleTextSize.SetRDefault("NULL")
+        ucrInputStyleTextSize.SetLinkedDisplayControl(lblStyleTextSize)
+
+
```
[the remark this answers is not in the document]
Comment (Ivanluv) [reply instat/sdgFormatSummaryTables.vb]: Done

File: instat/sdgFormatSummaryTables.vb (hunk 1 above)
Comment (Ivanluv) [reply instat/sdgFormatSummaryTables.vb]: Thank you for noticing this.Done

**Comment by rdstern** [comment]
@Ivanluv and possibly @lilyclements I have a curious feature in the new summary tables - which seems to be coming on fine.  And I am excited by my latest discovery - so let me start with that!
a) I mentioned to Ivan recently that I am keen for the Describe > One Variable > Summarise dialogue to be improved.  I have also found an issue with that dialogue when I ask for proportions.  It gives missing values and I seem to remember that this also happened - and was corrected - in this summary dialogue.
b) Then I found one of the items I wanted what to be able to change the orientation of the summaries in the one variable dialogue.  And of course, this is possible already in this summary dialogue!
c) I think you can see where this is all heading and (for me) it is simple and exciting!  The one variable dialogue (Customised) IS the summary dialogue, made simple - and with no factors!!!

So, back to the summary dialogue itself.  After that exciting discovery I used the summary dialogue with the recent R gui comparison data table that has been circulated.  Read 43 rows, so not the total at the bottom.  Then use your new summary dialogue.  I get all the results I want, but with 2 oddities!  Here is the dialogue and the output:

image

Here are the results:
image

However, they are not true!  It is like the Morecambe and Wise thing = "all the numbers are there, but not necessarily in the right order! " Both sets of names seem to be alphabetical, but it has really shuffled the columns I think!

The default is correct, (untick, both summaries as a factor and displays as rows) but then everything is in one very long row.  I think all other combinations add a shuffle!

This is a bit urgent to fix.  I should check it in 7.3.1!

**Review by rdstern — CHANGES_REQUESTED** [review]
@Ivanluv thanks for all those changes.  Just a couple of cosmetic changes.
a) It now says Display Outer and that should go back to Display Outer Margins.  Then you could delete the Margin from Margin Name, so there is space.
b) You have shortened the Factors and Variates controls.  I was thinking of leaving those and shortening the Variables control instead.  I think that will still be ok with the space.
c) And perhaps make the group box for the margins a little less generous, so it is still just a bit smaller overall?
d) Then there is the error we found to fix.  
e) then I hope @lloyddewit or @shadrackkibet can check the code soon, so we can merge.

**Comment by Ivanluv** [comment]
Am closing this PR due to the many conflicts

**Pull request closed without merging**